dev-tools/mage/crossbuild.go

```diff
@@ -0,0 +1,240 @@
+		if params.Serial {
+			if err := builder.Build(); err != nil {
+				return errors.Wrapf(err, "failed cross-building target=%v for platform=%v",
+					params.Target, buildPlatform.Name)
+			}
+		} else {
+			deps = append(deps, builder.Build)
+		}
+	}
+
+	// Each build runs in parallel.
+	Parallel(deps...)
+	return nil
+}
+
+// buildMage pre-compiles the magefile to a binary using the native GOOS/GOARCH
+// values for Docker. This is required to so that we can later pass GOOS and
+// GOARCH to mage for the cross-build. It has the benefit of speeding up the
+// build because the mage -compile is done only once rather than in each Docker
+// container.
+func buildMage() error {
+	env := map[string]string{
+		"GOOS":   "linux",
+		"GOARCH": "amd64",
+	}
+	return sh.RunWith(env, "mage", "-f", "-compile", filepath.Join("build", "mage-linux-amd64"))
+}
+
+func crossBuildImage(platform string) (string, error) {
+	tagSuffix := "main"
+
+	switch {
+	case strings.HasPrefix(platform, "darwin"):
+		tagSuffix = "darwin"
+	case strings.HasPrefix(platform, "linux/arm"):
+		tagSuffix = "arm"
+	case strings.HasPrefix(platform, "linux/mips"):
+		tagSuffix = "mips"
+	case strings.HasPrefix(platform, "linux/ppc"):
+		tagSuffix = "ppc"
+	case platform == "linux/s390x":
+		tagSuffix = "s390x"
+	case strings.HasPrefix(platform, "linux"):
+		// Use an older version of libc to gain greater OS compatibility.
+		// Debian 7 uses glibc 2.13.
+		tagSuffix = "main-debian7"
+	}
+
+	goVersion, err := GoVersion()
+	if err != nil {
+		return "", err
+	}
+
+	return beatsCrossBuildImage + ":" + goVersion + "-" + tagSuffix, nil
+}
+
+// GolangCrossBuilder executes the specified mage target inside of the
+// associated golang-crossbuild container image for the platform.
+type GolangCrossBuilder struct {
+	Platform string
+	Target   string
+}
+
+// Build executes the build inside of Docker.
+func (b GolangCrossBuilder) Build() error {
+	fmt.Printf(">> %v: Building for %v\n", b.Target, b.Platform)
+
+	repoInfo, err := GetProjectRepoInfo()
+	if err != nil {
+		return errors.Wrap(err, "failed to determine repo root and package sub dir")
+	}
+
+	mountPoint := filepath.ToSlash(filepath.Join("/go", "src", repoInfo.RootImportPath))
+	workDir := mountPoint
+	if repoInfo.SubDir != "" {
+		workDir = filepath.ToSlash(filepath.Join(workDir, repoInfo.SubDir))
+	}
+
+	dockerRun := sh.RunCmd("docker", "run")
+	image, err := crossBuildImage(b.Platform)
+	if err != nil {
+		return errors.Wrap(err, "failed to determine golang-crossbuild image tag")
+	}
+	verbose := ""
+	if mg.Verbose() {
+		verbose = "true"
+	}
+	var args []string
+	if runtime.GOOS != "windows" {
+		args = append(args,
+			"--env", "EXEC_UID="+strconv.Itoa(os.Getuid()),
+			"--env", "EXEC_GID="+strconv.Itoa(os.Getgid()),
+		)
+	}
+	args = append(args,
+		"--rm",
+		"--env", "MAGEFILE_VERBOSE="+verbose,
+		"--env", "MAGEFILE_TIMEOUT="+EnvOr("MAGEFILE_TIMEOUT", ""),
+		"--env", "CGO_ENABLED=1",
```

Comment:
Should this depend on the environment variable?

Reply:
It doesn't matter that it's set (or not set) here because the decision to use cgo for cross-builds is made by the `mage golangCrossBuild`. I'll remove this to avoid confusion.